### PR TITLE
Drag and Drop functionality

### PR DESCRIPTION
**Pull request by pavi41**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 1]
```diff
@@ -87,38 +121,61 @@ const FileUpload = (props: PROPS) => {
...
+            {files &&
+              files?.map((item: FileObject, index) => (
+                <li
+                  className="cmp-adaptiveform-fileinput__fileitem"
+                  key={item?.name}
+                >
+                  <span className="cmp-adaptiveform-fileinput__filename">
+                    {item?.name}
+                  </span>
+                  <span
```
Comment (rajatofficial) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:164]: add file size compatibility as per the core component.

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 2]
```diff
@@ -52,33 +67,52 @@ const FileUpload = (props: PROPS) => {
     return localFiles;
   }, []);
 
-  const handleNewFileUpload = useCallback((e: { target: { files: FileList | null; }; }) => {
-    const { files: newFiles } = e.target;
-    if (newFiles?.length) {
-      const updatedFiles = files.concat(fileListToArray(newFiles));
+  const handleDragOver = (event: React.DragEvent<HTMLDivElement>) => {
```
Comment (rajatofficial) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:70]: I don't think this function is required.

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 3]
```diff
@@ -87,38 +121,61 @@ const FileUpload = (props: PROPS) => {
...
+          >
+            <div className="cmp-adaptiveform-fileinput__icon"></div>
+            <div className="cmp-adaptiveform-fileinput__dragtext">
+              {properties?.dragDropText}
+            </div>
+            <label
+              htmlFor={id}
+              className="cmp-adaptiveform-fileinput__widgetlabel"
+            >
+              Attach
```
Comment (rajatofficial) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:138]: Why this 'Attach' text is hard coded it must be bind to json value?

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 2 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:70]: This is required. If it is removed the file gets open in other tab.

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 3 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:138]: The button text is not received in json

File: packages/react-vanilla-components/__tests__/components/FileUpload.test.tsx [hunk 4]
```diff
@@ -20,6 +22,9 @@ const field = {
   maxFileSize: "5MB",
   fieldType: "file-input",
   visible: true,
+  properties: {
+    dragDropText: "Drag and drop to Upload"
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/__tests__/components/FileUpload.test.tsx:26]: It should be "fd:dragDropText"

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 5]
```diff
@@ -26,20 +26,35 @@ import FieldWrapper from './common/FieldWrapper';
...
   let newMultiple = props.type?.endsWith('[]') ? { multiple: true } : {};
 
-  const callUpdateFilesCb = useCallback((files: Array<File | FileObject>) => {
-    if (newMultiple) {
-      props.dispatchChange(files);
-    } else {
-      props.dispatchChange(files.length > 0 ? files[0] : null);
-    }
-  }, [newMultiple, props.dispatchChange]);
+  const callUpdateFilesCb = useCallback(
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:48]: function name  should be as per functionality

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 6]
```diff
@@ -26,20 +26,35 @@ import FieldWrapper from './common/FieldWrapper';
...
+    visible,
+    enabled,
+    appliedCssClassNames,
+    properties
+  } = props;
+  let val = value && (value instanceof Array ? value : [value]);
   const [files, setFiles] = useState<FileObject[]>(val || []);
 
   const maxFileSizeInBytes = getFileSizeInBytes(maxFileSize);
   let newMultiple = props.type?.endsWith('[]') ? { multiple: true } : {};
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:46]: rename the variable name.

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 7]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
     return localFiles;
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:67]: ?

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 8]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
   }, []);
 
-  const handleNewFileUpload = useCallback((e: { target: { files: FileList | null; }; }) => {
-    const { files: newFiles } = e.target;
-    if (newFiles?.length) {
-      const updatedFiles = files.concat(fileListToArray(newFiles));
+  const handleDragOver = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    const tgt = event.target as Element;
+    console.log(tgt);
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:73]: remove console

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 9]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+    tgt.classList.add('cmp-adaptiveform-fileinput__dragarea--active');
+  };
+
+  const handleDragLeave = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    const tgt = event.target as Element;
+    tgt.classList.remove('cmp-adaptiveform-fileinput__dragarea--active');
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:80]: add/remove class based on the state. Don’t use the js selector in react.

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 10]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+  };
+
+  const handleNewFileUpload = useCallback(
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:83]: rename the function name with this "fileUploadHandler".

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 11]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+    (
+      e:
+        | { target: { files: FileList | null } }
+        | React.DragEvent<HTMLDivElement>
+    ) => {
+      //@ts-ignore
+      e.preventDefault();
+      //@ts-ignore
+      const { files: newFiles } = e.dataTransfer ? e.dataTransfer : e.target;
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:92]: In which case, files are not available in the target object

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 12]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+  const handleDragLeave = (event: React.DragEvent<HTMLDivElement>) => {
+    event.preventDefault();
+    const tgt = event.target as Element;
+    tgt.classList.remove('cmp-adaptiveform-fileinput__dragarea--active');
+  };
+
+  const handleNewFileUpload = useCallback(
+    (
+      e:
+        | { target: { files: FileList | null } }
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:86]: can you simplify this?

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 13]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+      }
+      const tgt = e.target as Element;
+      tgt.classList.remove('cmp-adaptiveform-fileinput__dragarea--active');
+    },
+    [setFiles, callUpdateFilesCb]
+  );
+
+  const removeFile = useCallback(
+    (index: number) => {
+      const updatedFiles = files.slice(0, index).concat(files.slice(index + 1));
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:106]: you can remove file like this - 
files.splice(0, index)

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 14]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
       setFiles(updatedFiles);
       callUpdateFilesCb(updatedFiles);
+    },
+    [callUpdateFilesCb, setFiles]
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:110]: add files dependencies

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 15]
```diff
@@ -52,33 +67,75 @@ const FileUpload = (props: PROPS) => {
...
+  );
+
+  const formatBytes = (bytes: number, decimals = 0) => {
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:113]: move this method in utils files

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 16]
```diff
@@ -87,38 +144,69 @@ const FileUpload = (props: PROPS) => {
...
+            </label>
+            <input
+              ref={fileInputField}
+              className="cmp-adaptiveform-fileinput__widget"
+              id={id}
+              type="file"
+              name={name}
+              onChange={handleNewFileUpload}
+              required={required}
+              accept={accept?.toString()}
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:172]: accept can be undefined

File: packages/react-vanilla-components/src/components/FileUpload.tsx [hunk 17]
```diff
@@ -87,38 +144,69 @@ const FileUpload = (props: PROPS) => {
...
+                    className="cmp-adaptiveform-fileinput__filename"
+                    aria-label={item?.name}
+                  >
+                    {item?.name}
+                  </span>
+                  <span className="cmp-adaptiveform-fileinput__fileendcontainer">
+                    <span className="cmp-adaptiveform-fileinput__filesize">
+                      {formatBytes(item?.size)}
+                    </span>
+                    <span
```
Comment (amit-mnnit) [inline packages/react-vanilla-components/src/components/FileUpload.tsx:195]: please use button not span atg

**Review by amit-mnnit — CHANGES_REQUESTED** [review]
Please check comments

File: packages/react-vanilla-components/__tests__/components/FileUpload.test.tsx (hunk 4 above)
Comment (pavi41) [reply packages/react-vanilla-components/__tests__/components/FileUpload.test.tsx:26]: Fetching this from JSON. Correction needs to be done in core component

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 8 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:73]: Removed!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 10 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:83]: Done!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 9 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:80]: Implemented!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 17 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:195]: Implemented!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 6 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:46]: Done!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 12 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:86]: Done!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 15 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx:113]: Moved!

File: packages/react-vanilla-components/src/components/FileUpload.tsx (hunk 5 above)
Comment (pavi41) [reply packages/react-vanilla-components/src/components/FileUpload.tsx]: Changed!

**Review by amit-mnnit — APPROVED** [review]

**Pull request merged**